internal/backend/fs.go

```diff
@@ -152,26 +162,66 @@ func (s *storageFS) CreateObject(obj Object) (Object, error) {
 	defer s.mtx.Unlock()
 	err := s.createBucket(obj.BucketName)
 	if err != nil {
-		return Object{}, err
+		return StreamingObject{}, err
 	}
 
 	path := filepath.Join(s.rootDir, url.PathEscape(obj.BucketName), url.PathEscape(obj.Name))
 
-	if err = os.WriteFile(path, obj.Content, 0o600); err != nil {
-		return Object{}, err
+	tempFile, err := os.CreateTemp("", "fake-gcs-object")
+	if err != nil {
+		return StreamingObject{}, err
 	}
+	tempFile.Close()
+
+	tempFile, err = compatOpenForWritingAllowingRename(tempFile.Name())
+	if err != nil {
+		return StreamingObject{}, err
+	}
+	defer tempFile.Close()
+
+	// The file is renamed below, which causes this to be a no-op. If the
+	// function returns before the rename, though, the temp file will be
+	// removed.
+	defer os.RemoveAll(tempFile.Name())
```

Comment:
nit: os.Remove should be enough?
```suggestion
	defer os.Remove(tempFile.Name())
```